docs: enhance database operations examples with practical use cases

Vallabh-Mahanta opened the pull request:
This PR enhances the database operations documentation with more practical examples including:
   
   - Detailed CRUD operations for products with realistic field values
   - Custom repository examples with specialized query methods
   - Complex transaction examples showing multiple operations in a single atomic transaction
   - Advanced SQL query examples for common e-commerce analytics needs
   
   These improvements will help developers better understand how to effectively work with the database in Medusa v2.

Commit:
docs: add guide for database access in v2 (fixes #12216)

Commit:
docs: enhance database operations examples with practical use cases

shahednasser commented on www/apps/book/app/learn/fundamentals/modules/db-operations/page.mdx:
````diff
@@ -353,7 +353,78 @@ You can then use the data model repository in your service to perform database o
+Here are examples of typical operations using a repository:
+
+```ts
+// Create a new product
+const product = await productRepository.create({
+  title: "Ergonomic Chair",
+  description: "Office chair with lumbar support",
+  price: 299.99,
+  inventory_quantity: 50
+})
+
+// Find a specific product
+const product = await productRepository.findOne({
+  where: { id: "prod_123456789" }
+})
+
+// Update a product
+await productRepository.update("prod_123456789", {
+  inventory_quantity: 42,
+  metadata: {
+    featured: true
+  }
+})
+
+// Delete a product
+await productRepository.delete("prod_123456789")
+
+// Find multiple products with filtering
+const products = await productRepository.find({
+  where: { 
+    status: "published", 
+    inventory_quantity: { $gt: 0 }
+  },
+  take: 10,
+  skip: 0,
+  order: { created_at: "DESC" }
+})
````
Since the document constantly refers to a `post` model, can you use `postReopsitory` examples instead? You can use fields relevant to posts like `title` or `published_at`.

shahednasser replied:
The pagination fields in `find` are incorrect. Refer to the method's reference for more details: https://docs.medusajs.com/resources/data-model-repository-reference/methods/find

shahednasser commented on www/apps/book/app/learn/fundamentals/modules/db-operations/page.mdx:
```diff
@@ -353,7 +353,78 @@ You can then use the data model repository in your service to perform database o
+  inventory_quantity: 50
+})
+
+// Find a specific product
+const product = await productRepository.findOne({
+  where: { id: "prod_123456789" }
+})
```
This isn't available in a data model repository, only with the entity manager, so would remove it from here.

shahednasser commented on www/apps/book/app/learn/fundamentals/modules/db-operations/page.mdx:
```diff
@@ -353,7 +353,78 @@ You can then use the data model repository in your service to perform database o
+})
+
+// Update a product
+await productRepository.update("prod_123456789", {
+  inventory_quantity: 42,
+  metadata: {
+    featured: true
+  }
+})
```
The `update` usage is incorrect, refer to the update reference for more details: https://docs.medusajs.com/resources/data-model-repository-reference/methods/update

shahednasser commented on www/apps/book/app/learn/fundamentals/modules/db-operations/page.mdx:
```diff
@@ -353,7 +353,78 @@ You can then use the data model repository in your service to perform database o
+})
+
+// Delete a product
+await productRepository.delete("prod_123456789")
```
The `delete` usage is incorrect, please refer to the delete reference for more details: https://docs.medusajs.com/resources/data-model-repository-reference/methods/delete

shahednasser commented on www/apps/book/app/learn/fundamentals/modules/db-operations/page.mdx:
````diff
@@ -353,7 +353,78 @@ You can then use the data model repository in your service to perform database o
+})
+```
+
+You can also create custom repository classes with specialized query methods:
+
+```ts
+import { DALUtils } from "@medusajs/framework/utils"
+import { Product } from "./models/product"
+
+export class ProductRepository extends DALUtils.mikroOrmBaseRepositoryFactory(Product) {
+  // Add custom methods for specialized queries
+  async findAvailableProducts() {
+    return await this.find({
+      where: { status: "published", inventory_quantity: { $gt: 0 } }
+    })
+  }
+  
+  async findFeaturedProducts() {
+    return await this.find({
+      where: { 
+        metadata: {
+          featured: true
+        }
+      }
+    })
+  }
+  
+  async updateInventory(productId: string, quantity: number) {
+    return await this.update(productId, {
+      inventory_quantity: quantity
+    })
+  }
+}
+```
````
Please remove this as the main recommended approach at the moment is to either use the inject manager or the data model repository.

shahednasser commented on www/apps/book/app/learn/fundamentals/modules/db-operations/page.mdx:
````diff
@@ -683,3 +754,74 @@ class BlogModuleService {
+```ts
+// Get popular products with their categories
+async function getPopularProducts() {
+  const knex = this.manager_.getKnex()
+  
+  return await knex.raw(`
+    SELECT 
+      p.id, 
+      p.title, 
+      p.price, 
+      c.name as category_name,
+      COUNT(oi.id) as order_count
+    FROM products p
+    LEFT JOIN product_categories pc ON p.id = pc.product_id
+    LEFT JOIN categories c ON pc.category_id = c.id
+    LEFT JOIN order_items oi ON p.id = oi.product_id
+    GROUP BY p.id, c.id
+    ORDER BY order_count DESC
+    LIMIT 10
+  `)
+}
+
+// Calculate sales metrics
+async function getSalesMetrics() {
+  const knex = this.manager_.getKnex()
+  
+  return await knex.raw(`
+    SELECT 
+      DATE_TRUNC('month', o.created_at) as month,
+      SUM(o.total) as revenue,
+      COUNT(o.id) as orders,
+      COUNT(DISTINCT o.customer_id) as unique_customers
+    FROM orders o
+    WHERE o.status = 'completed'
+    GROUP BY month
+    ORDER BY month DESC
+    LIMIT 12
+  `)
+}
+```
````
The practices in these code snippets don't follow our best practices, as they try to create / query data from other modules (which violates module isolation). If you want to add examples, please use the practices recommended in this documentation page.

shahednasser commented on www/apps/docs/app/modules/database-access/page.mdx:
Remove page please

Vallabh-Mahanta replied:
I am sorry for this, I am removing it, I knew I made a mistake while it was checking so I was waiting once the checking gets completed.

shahednasser reviewed (changes requested):
Thanks for your contribution!

Vallabh-Mahanta committed:
Delete www/apps/docs/app/modules/database-access/page.mdx

Added this by mistake, please excuse me.

Vallabh-Mahanta committed:
Update page.mdx

I am sorry for this, I thought for examples we can give any which make others to understand the code more diversely, Now made the changes according to modules, I hope this meets all the requirements.

Vallabh-Mahanta reviewed (commented):
I am sorry I didn't noticed, I thought for examples we can give any which makes others to understand the code more diversely, I hope this meet all the requirements.

shahednasser commented on www/apps/book/app/learn/fundamentals/modules/db-operations/page.mdx:
```diff
@@ -683,3 +683,146 @@ class BlogModuleService {
+  status: "draft"
+})
+
+// Find a specific post
+const post = await postRepository.findOne({
+  where: { id: "post_123456" }
+})
```
The `findOne` method isn't available on a data model repository so remove it please.

shahednasser commented on www/apps/book/app/learn/fundamentals/modules/db-operations/page.mdx:
```diff
@@ -683,3 +683,146 @@ class BlogModuleService {
+})
+
+// Update a post
+await postRepository.update("post_123456", {
+  status: "published",
+  metadata: {
+    featured: true
+  }
+})
```
Again you need to update it to match the update reference: https://docs.medusajs.com/resources/data-model-repository-reference/methods/update

shahednasser commented on www/apps/book/app/learn/fundamentals/modules/db-operations/page.mdx:
```diff
@@ -683,3 +683,146 @@ class BlogModuleService {
+})
+
+// Delete a post
+await postRepository.delete("post_123456")
```
Needs an update to match the delete reference: https://docs.medusajs.com/resources/data-model-repository-reference/methods/delete

shahednasser commented on www/apps/book/app/learn/fundamentals/modules/db-operations/page.mdx:
```diff
@@ -683,3 +683,146 @@ class BlogModuleService {
+await postRepository.delete("post_123456")
+
+// Find multiple posts with filtering
+const posts = await postRepository.find({
+  where: { 
+    status: "published", 
+    author_id: "author_123"
+  },
+  take: 10,
+  skip: 0,
+  order: { created_at: "DESC" }
+})
```
Need an update to match the find reference: https://docs.medusajs.com/resources/data-model-repository-reference/methods/find

shahednasser commented on www/apps/book/app/learn/fundamentals/modules/db-operations/page.mdx:
````diff
@@ -683,3 +683,146 @@ class BlogModuleService {
+})
+```
+
+You can also create custom repository classes with specialized query methods:
+
+```ts
+import { DALUtils } from "@medusajs/framework/utils"
+import { Post } from "./models/post"
+
+export class PostRepository extends DALUtils.mikroOrmBaseRepositoryFactory(Post) {
+  // Add custom methods for specialized queries
+  async findPublishedPosts() {
+    return await this.find({
+      where: { status: "published" }
+    })
+  }
+  
+  async findFeaturedPosts() {
+    return await this.find({
+      where: { 
+        metadata: {
+          featured: true
+        }
+      }
+    })
+  }
+  
+  async updateStatus(postId: string, status: string) {
+    return await this.update(postId, {
+      status
+    })
+  }
+}
+```
````
Remove this as mentioned before.

shahednasser commented on www/apps/book/app/learn/fundamentals/modules/db-operations/page.mdx:
````diff
@@ -683,3 +683,146 @@ class BlogModuleService {
+```ts
+// Get popular posts with their categories
+async function getPopularPosts() {
+  const knex = this.manager_.getKnex()
+  
+  return await knex.raw(`
+    SELECT 
+      p.id, 
+      p.title, 
+      p.created_at, 
+      c.name as category_name,
+      COUNT(cm.id) as comment_count
+    FROM posts p
+    LEFT JOIN post_categories pc ON p.id = pc.post_id
+    LEFT JOIN categories c ON pc.category_id = c.id
+    LEFT JOIN comments cm ON p.id = cm.post_id
+    WHERE p.status = 'published'
+    GROUP BY p.id, c.id
+    ORDER BY comment_count DESC
+    LIMIT 10
+  `)
+}
+
+// Calculate blog metrics
+async function getBlogMetrics() {
+  const knex = this.manager_.getKnex()
+  
+  return await knex.raw(`
+    SELECT 
+      DATE_TRUNC('month', p.created_at) as month,
+      COUNT(p.id) as posts,
+      COUNT(DISTINCT p.author_id) as unique_authors,
+      SUM(CASE WHEN p.status = 'published' THEN 1 ELSE 0 END) as published_posts
+    FROM posts p
+    GROUP BY month
+    ORDER BY month DESC
+    LIMIT 12
+  `)
+}
+```
````
Use the entity manager as shown in the first section:

https://docs.medusajs.com/learn/fundamentals/modules/db-operations#run-queries

Also it would make sense to move this part into the Run Queries section.

Vallabh-Mahanta committed:
Update page.mdx

I have made some changes after going through the documentations you have linked and I am sorry for leaving this in middle, I had some academic tests. But now I am completely free from it and now I will be actively contributing to this repo. Also thank you for the patience.

Vallabh-Mahanta reviewed (commented):
Please review it I have done some changes and sorry for late I was busy with some of my academics but I will now actively contributing to this repo. Please review the changes I made.

shahednasser reviewed (changes requested):
Thanks for your contribution!

We already have a [reference](https://docs.medusajs.com/resources/data-model-repository-reference) that covers all the methods of a data model repository, like `find` an `update`, and we link to it in the earlier section about data model repositories. As for the execute example, I'm fine with keeping it if we move it to the `Run Queries` section with a header like "Complex Database Query Example"

Vallabh-Mahanta committed:
Update page.mdx

These patterns adhere to Medusa's best practices for:

1. Transaction boundary management
2. Query optimization through raw SQL execution
3. Scalable analytical processing
4. Maintainable service layer architecture

Vallabh-Mahanta committed:
Update page.mdx

These patterns adhere to Medusa's best practices for:

1. Transaction boundary management
2. Query optimization through raw SQL execution
3. Scalable analytical processing
4. Maintainable service layer architecture

Vallabh-Mahanta reviewed (commented):
@shahednasser ma'am before earlier as I have gave examples for find, delete, update etc, were the extra examples which you have asked me to implement in this docs or may be I misunderstood. Now I have replaced them all and kept them under Run Queries section which you wanted me to do.

shahednasser reviewed (changes requested):
Sorry for the confusion, but your PR was opened a while ago. Now we have the data repository reference, so the examples for find, update, etc... are not necessary

The change needed right now is to remove those examples, and only move the execute examples to be under [this section](https://docs.medusajs.com/learn/fundamentals/modules/db-operations#run-queries)

Vallabh-Mahanta committed:
Update page.mdx

The changes maintain:
- Consistent heading hierarchy
- Transaction management patterns using @InjectTransactionManager
- MikroORM entity manager usage
- Standard documentation components

Vallabh-Mahanta reviewed (commented):
@shahednasser Apologies for the delay in opening the PR. I’ve been busy with my academic exams, but I’ve now completed them. I’ll be fully contributing to this repo moving forward.
